app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java

```
@@ -1,40 +1,54 @@
+        {
+            put(AllowInvisibleElements.SETTING_NAME, AllowInvisibleElements.class);
+            put(CompressedLayoutHierarchy.SETTING_NAME, CompressedLayoutHierarchy.class);
+            put(WaitForIdleTimeout.SETTING_NAME, WaitForIdleTimeout.class);
+            put(EnableNotificationListener.SETTING_NAME, EnableNotificationListener.class);
+        }
+    };
+
     public UpdateSettings(String mappedUri) {
         super(mappedUri);
     }
 
     @Override
     public AppiumResponse safeHandle(IHttpRequest request) {
-
         try {
             Map<String, Object> settings = getPayload(request, "settings");
-            Session.capabilities.putAll(settings);
             Logger.debug("Update settings: " + settings.toString());
-
-            if (settings.containsKey(AllowInvisibleElements.SETTING_NAME)) {
-                AllowInvisibleElements.updateSetting((boolean) settings.get(AllowInvisibleElements.SETTING_NAME));
-            }
-
-            if (settings.containsKey(CompressedLayoutHierarchy.SETTING_NAME)) {
-                CompressedLayoutHierarchy.updateSetting((boolean) settings.get(CompressedLayoutHierarchy.SETTING_NAME));
-            }
-
-            if (settings.containsKey(WaitForIdleTimeout.SETTING_NAME)) {
-                WaitForIdleTimeout.updateSetting((int) settings.get(WaitForIdleTimeout.SETTING_NAME));
+            for (Entry<String, Object> entry : settings.entrySet()) {
+                String settingName = entry.getKey();
+                Object settingValue = entry.getValue();
+                ISetting setting = getSetting(settingName);
+                if (setting != null) {
+                    setting.updateSetting(settingValue);
+                    Session.capabilities.put(settingName, settingValue);
+                } else {
+                    Logger.error(String.format("Setting '%s' is not supported.", entry.getKey()));
```

shall we throw an exception instead?

---

From my point of view we can just skip unsupported setting like unsupported desired capabilities.

---

unsupported caps are visible in the server log (and even there not so many people notice this). How do you think - how many of them are going to check device logs?

---

ok

---

This all goes through the driver, so validation/logging can happen there. Or the calling code could catch an error and log it, while continuing to function normally.